packages/react-utils/src/components/FhirSelect/index.tsx

```diff
@@ -0,0 +1,169 @@
+import React, { useEffect, useState } from 'react';
+import { URLParams } from '@opensrp/server-service';
+import { IBundle } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IBundle';
+import { useInfiniteQuery } from 'react-query';
+import { PlusOutlined } from '@ant-design/icons';
+import { Button, Divider, Select, Empty, Space, Spin, Alert } from 'antd';
+import type { SelectProps } from 'antd';
+import { IResource } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IResource';
+import { debounce } from 'lodash';
+import { getResourcesFromBundle } from '../../helpers/utils';
+import { useTranslation } from '../../mls';
+import { loadResources, getTotalRecordsInBundles, getTotalRecordsOnApi } from './utils';
+
+type SelectOption<T extends IResource> = {
+  label: string;
+  value: string | number;
+  ref: T;
+};
+
+export interface TransformOptions<T extends IResource> {
+  (resource: T): SelectOption<T> | undefined;
+}
+
+export type AbstractedSelectOptions<ResourceT extends IResource> = Omit<
+  SelectProps<string, SelectOption<ResourceT>>,
+  'loading' | 'options' | 'searchValue'
+>;
+
+export interface FhirSelectProps<ResourceT extends IResource>
+  extends AbstractedSelectOptions<ResourceT> {
+  resourceType: string;
+  baseUrl: string;
+  transformOption: TransformOptions<ResourceT>;
+  filterPageSize?: number;
+  extraQueryParams?: URLParams;
+  getFullOptionOnChange?: (obj: SelectOption<ResourceT> | SelectOption<ResourceT>[]) => void;
+}
+
+const debouncedFn = debounce((callback) => callback(), 500);
+
+/**
+  Select component to use with async search
+ *
+ * @param props - component props
+ */
+export function FhirSelect<ResourceT extends IResource>(props: FhirSelectProps<ResourceT>) {
+  const {
+    resourceType,
+    baseUrl,
+    transformOption,
+    placeholder,
+    filterPageSize: pageSize = 20,
+    extraQueryParams = {},
+    getFullOptionOnChange,
+    ...restProps
+  } = props;
+  const defaultStartPage = 1;
+  const [page, setPage] = useState(defaultStartPage);
+  const [searchValue, setSearchValue] = React.useState<string>();
+  const [debouncedSearchValue, setDebouncedSearchValue] = React.useState<string>();
+  const { t } = useTranslation();
+
+  useEffect(() => {
```

Comment:
hmm,.. What would be the advantage in that case.

Reply:
re-render's only on dependency change instead of after every x debounce period

Reply:
The reason I have the debounce function is because the inifinity query uses the search value as a query key. Debouncing the searchValue helps prevent the infinity query from triggering everytime there is an input during search.
I however think this can be optimized by adding the searchValue as a dependency array, but it still won't remove the need for debouncing the search value.